tests/base/cart.spec.ts

```diff
@@ -50,7 +58,70 @@ test.describe('Cart functionalities', () => {
     await cart.removeProduct(selectors.productPage.simpleProductTitle);
   });
 
-});
 
-//TODO: Write test to add coupon
-//TODO: Write test to remove coupon
+  /**
+   * @feature Discount Code
+   * @scenario User adds a discount code to their cart
+   * @given I have a product in my cart
+   *  @and I am on my cart page
+   * @when I click on the 'add discount code' button
+   * @then I fill in a code
+   *  @and I click on 'apply code'
+   * @then I should see a confirmation that my code has been added
+   *  @and the code should be visible in the cart
+   *  @and a discount should be applied to the product
+   */
+  test('Add coupon code in cart',{ tag: ['@cart', '@coupon-code']}, async ({page}) => {
+    const cart = new CartPage(page);
+    let discountCode = process.env.DISCOUNT_CODE;
+
+    if(!discountCode) {
+      throw new Error(`discountCode appears to not be set in .env file. Value reported: ${discountCode}`);
+    }
+
+    await cart.applyDiscountCode(discountCode);
+  });
+
+  /**
+   * @feature Remove discount code from cart
+   * @scenario User has added a discount code, then removes it
+   * @given I have a product in my cart
+   * @and I am on my cart page
+   * @when I add a discount code
+   * @then I should see a notification
+   * @and the code should be visible in the cart
+   * @and a discount should be applied to a product
+   * @when I click the 'cancel coupon' button
+   * @then I should see a notification the discount has been removed
+   * @and the discount should no longer be visible.
+   */
+  test('Remove coupon code from cart',{ tag: ['@cart', '@coupon-code'] }, async ({page}) => {
+    const cart = new CartPage(page);
+    let discountCode = process.env.DISCOUNT_CODE;
```

Comment:
@shayfaber See comments above. Same for this code, maybe also an idea to move this code to a function, because you have the same code twice now.